tools/kann

```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
+
+set -e
+pushd $(cd $(dirname ${0})/..; pwd) > /dev/null
+
+third_party_dir="./third_party/source/kann"
+
+clean() {
+    echo "kann clean"
+    rm -rf "${third_party_dir}"
+}
+
+sync() {
+    echo "kann sync"
+    mkdir -p "${third_party_dir}/metadata"
+    curl --silent --location --output "${third_party_dir}/metadata/kann.fbs" "https://github.com/kalray/kann-models-zoo/raw/refs/heads/main/utils/third-party/kann.fbs"
+}
+
+# regenerate kann-schema.js with 'tools/kann schema'
+schema() {
+    echo "kann schema"
+    [[ $(grep -U $'\x0D' ./source/kann-schema.js) ]] && crlf=1
```

Comment:
@zylfried running `./tools/kann sync schema` creates an invalid file, missing the namespace? Instead of a `third_party` folder in the model zoo can the `.fbs` file be move to a more official folder or repo with some format documentation.

Reply:
Indeed the namespace was missing.
The thing about this `.fbs` file is that the format used for Netron isn't the final official serialized format as explained in previous answers. Since it's an intermediate serialization for now used for Netron visualization, keeping it in `third_party` seems appropriate. The official format is documented in our main repo.

Reply:
@zylfried can you elaborate and share a link to the previous answer you are referring to as well as a link to the documentation for the official format? Why is there a different serialization format instead of using the official one? Is this format only used for visualization purposes?

Reply:
Yes this is for now only used for visualization, here are some of my previous answers :

> It's important to clarify that KaNN is an inference engine similar to TensorRT, so the graphs we handle are more low-level than ONNX models. We have our own internal model format, and we'd like to use Netron for visualizing it but the difference in structure and purpose makes adapting existing formats insufficient to meet KaNN’s specific needs.

> KaNN’s purpose is not to introduce yet another framework for neural network design but rather to translate models from popular frameworks into a lower-level, processor-executable format. The `.kann` graph we aim to visualize in Netron represents an intermediate stage of model optimization and refactorization. This stage is then compiled into processor-level kernels, which is the final usable format but could not be visualized in Netron, the intermediate format being therefore used for visualization purposes only.

Full documentation is available on the model zoo repo if you're interested in our work [https://github.com/kalray/kann-models-zoo](https://github.com/kalray/kann-models-zoo)

Reply:
@zylfried instead of introducing another custom format only for visualization, what about exploring exporting to one of the existing 50+ formats the app supports or working on improving the built-in message format.

Reply:
I thought we'd already answered this topic, plus `.kann` or `.kgraph` are not temporary and may grow in importance over time but for now we'd like to use them for visualization to improve our process.

Reply:
If the user workflow requires exporting to a special format for visualization it is not clear this warrants maintaining yet another format. The deep learning format space is already very fragmented with often redundant formats that deviate only in implementation details. It wasn't clear that users have to create these files orthogonal to the framework in the context of visualization which makes it even less desirable as supporting the format wouldn't help streamlining the user workflow. Instead it might make more sense to export to a standardized format that supports all capabilities of Netron and can be shared by other frameworks. For example, exporting to ONNX as a container or  standardizing on visualization metadata seems a good starting point. `.message` was intended to represent an object model that reflects Netron capabilities and is used by some users for this purpose but isn't fully developed.

Reply:
The current implementation of arcs is not supported in Netron. e.g. using string manipulation to represent information that Netron can't process is not the right approach. The question seems more if there is a common denominator. For example, can the graph be presented in another format. The rendered graph that looks like an ONNX graph in the current implementation, there aren't any restrictions which operators ONNX can represent. Can extra information be treaded as metadata. For example, the current use of shape is not an actual shape, should it be annotated as metadata instead as it is not a construct the tool understands. Given the format is only used temporarily and there are many redundant formats already, what options have you explained to avoid introducing and maintaining a new format.

Reply:
> nodes are then not whole tensors but subviews (smaller parts of the tensors). These subviews correspond to the arcs in our graph, with an Arc being a connection between nodes, as defined in graph theory

Is there more documentation available explaining the motivations behind this design. The current implementation seems to reduce this to strings and is misusing tensor shapes, is the implementation actually relevant?

Reply:
We completely understand your concern about introducing yet another format to maintain, and we have explored adapting ONNX. However, it does not meet our specific requirements. Unlike ONNX and other formats supported by Netron, KaNN focuses on managing data across processor memory rather than operators. Without going into too much detail, images may be split into multiple tensors, stored in different memory regions, and processed in parallel by operators. To represent this in the graph, we need to manipulate strings not fully supported otherwise. Full documentation of the internal architecture isn't publicly available at this time.
Does this address your concerns? We are willing to address all your requests, but certain constraints in our format do not make it always possible.

Reply:
Makes sense. Generally it should be possible to export to a format like `message.js` or complete the implementation to include all features of the app. What remains are unsupported features. For example, `subview` is not a supported type in the app and multi-line text concatenation for tensor types is not supported. Such constructs would have to be documented and natively supported in the app or removed from the implementation.

Reply:
Thank you for your feedback and support. We understand your perspective, but what you’re suggesting now would require restarting work we thought had your initial agreement. We explored Netron’s constraints following your instructions (https://github.com/lutzroeder/netron/issues/1308 and more) and concluded a new format was necessary based on prior discussions. If `message.js` is now the preferred approach, earlier guidance would have been helpful to avoid months of back-and-forth. At this stage, starting over isn’t feasible for us. What do you think about integrating our current proposal?